### PR TITLE
fix(stepfunctions-tasks): Athena StartQueryExecution includes QueryExecutionContext even when object is empty

### LINKED ISSUE #16133: (aws-stepfunctions-tasks): Athena QueryExecutionContext on StartQueryExecution is superfluous and leads to error
The following error appears in my stepfunction console:
```json
{
  "resourceType": "athena",
  "resource": "startQueryExecution.sync",
  "error": "Athena.InvalidRequestException",
  "cause": "Both queryExecutionContext.catalog and queryExecutionContext.database are null or empty (Service: AmazonAthena; Status Code: 400; Error Code: InvalidRequestException; Request ID: [Censored]; Proxy: null)"
}
```

### Reproduction Steps

I use the following code to start the Athena Query execution:

```ts
new AthenaStartQueryExecution(this,"stepfunction-start-query", {
            queryString: `[Some Query]`,
            integrationPattern: IntegrationPattern.RUN_JOB,
            resultConfiguration: {
                outputLocation: {
                    bucketName: bucket.bucketName,
                    objectKey: `States.Format('athena/{}/{}', $$.Execution.StartTime, $$.State.EnteredTime)`,
                },
            }
        })
```
apart from the not really working object Key (which is a problem I aim to resolve later on), the code seems to insert a queryExecutionContext in the generated State Machine.

I retrieved the following snippet from the stepfunction console:
```json
"stepfunction-start-query": {
      "Type": "Task",
      "Resource": "arn:aws:states:::athena:startQueryExecution.sync",
      "Parameters": {
        "QueryString": "[Some Query]",
        "QueryExecutionContext": {},
        "ResultConfiguration": {
          "OutputLocation": "s3://[Bucket]/States.Format('athena/{}/{}', $$.Execution.StartTime, $$.State.EnteredTime)/"
        }
      }
    }
```

### What did you expect to happen?

As `queryExecutionContext` is marked as optional in https://github.com/aws/aws-cdk/blob/v1.119.0/packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts#L30 and the comment states `@default - No query execution context` I did not expect CDK to generate any QueryExecutionContext.

### What actually happened?

In line https://github.com/aws/aws-cdk/blob/v1.119.0/packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts#L192 one can see, that only the values are being toggled, but not the whole QueryExecutionContext which is the source of the error.

### Environment

  - **CDK CLI Version  : Current main branch**

### Other

To fix this, I suggest on dymamically inserting the QueryExecutionContext when its defined (it cannot be null either) like this:

```ts
protected _renderTask(): any {

    let renderedObject = {
        Resource: integrationResourceArn('athena', 'startQueryExecution', this.integrationPattern),
        Parameters: sfn.FieldUtils.renderObject({
            QueryString: this.props.queryString,
            ClientRequestToken: this.props.clientRequestToken,
            ResultConfiguration: {
                EncryptionConfiguration: this.renderEncryption(),
            },
            WorkGroup: this.props.workGroup,
        }),
    }


    if (this.props.resultConfiguration?.outputLocation) {
        renderedObject.ResultConfiguration.OutputLocation = `s3://${this.props.resultConfiguration?.outputLocation?.bucketName}/${this.props.resultConfiguration?.outputLocation?.objectKey}/`,
    }

    if (this.props?.queryExecutionContext) {
        renderedObject.QueryExecutionContext = {
            Catalog: this.props.queryExecutionContext?.catalogName,
            Database: this.props.queryExecutionContext?.databaseName,
        }
    }


    return renderedObject;
}
```

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by LukvonStrom**
fixes: #16133

As described in #16133, the empty  `QueryExecutionContext`, which was included before this fix, leads to an error.
To mitigate this, I implemented a check to exclude the  `QueryExecutionContext`when both params are empty.
This way, it behaves like a truly optional parameter.

Additionally I integrated an easier way to merge an OutputLocation if provided.
This means less redundancy than the original solution, which returned an object that was only different in one key in each branch.

The unittest is based on the default unittest, but a bit "stripped down". It proofs that no empty `QueryExecutionContext` object is present.

In my tests, this fixes the 
```
{
  "resourceType": "athena",
  "resource": "startQueryExecution.sync",
  "error": "Athena.InvalidRequestException",
  "cause": "Both queryExecutionContext.catalog and queryExecutionContext.database are null or empty (Service: AmazonAthena; Status Code: 400; Error Code: InvalidRequestException; Request ID: [Censored]; Proxy: null)"
}
```
error.

This solution turned out to be a bit different than I envisioned in the issue.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 1]
```diff
@@ -183,40 +183,35 @@ export class AthenaStartQueryExecution extends sfn.TaskStateBase {
    * @internal
    */
   protected _renderTask(): any {
+
+    let renderedObject = {
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:187]: prefer using `const` over `let` unless specifically unnecessary
```suggestion
    const renderedObject = {
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 2]
```diff
@@ -183,40 +183,35 @@ export class AthenaStartQueryExecution extends sfn.TaskStateBase {
...
+      Resource: integrationResourceArn('athena', 'startQueryExecution', this.integrationPattern),
+      Parameters: {
+        QueryString: this.props.queryString,
+        ClientRequestToken: this.props.clientRequestToken,
+        QueryExecutionContext: {
+          Catalog: this.props.queryExecutionContext?.catalogName,
+          Database: this.props.queryExecutionContext?.databaseName,
+        },
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:195]:
```suggestion
        QueryExecutionContext: (this.props.queryExecutionContext?.catalogName || this.props.queryExecutionContext?.databaseName) ? {
          Catalog: this.props.queryExecutionContext?.catalogName,
          Database: this.props.queryExecutionContext?.databaseName,
        } : undefined,
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 3]
```diff
@@ -183,40 +183,35 @@ export class AthenaStartQueryExecution extends sfn.TaskStateBase {
...
+        ResultConfiguration: {
+          EncryptionConfiguration: this.renderEncryption(),
+        },
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:198]: let's skip the `if/else` altogether
```suggestion
          EncryptionConfiguration: this.renderEncryption(),
          OutputLocation: this.props.resultConfiguration?.outputLocation ? `s3://${this.props.resultConfiguration?.outputLocation?.bucketName}/${this.props.resultConfiguration?.outputLocation?.objectKey}/` : undefined,
        },
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 4]
```diff
@@ -183,40 +183,35 @@ export class AthenaStartQueryExecution extends sfn.TaskStateBase {
...
+        OutputLocation: `s3://${this.props.resultConfiguration?.outputLocation?.bucketName}/${this.props.resultConfiguration?.outputLocation?.objectKey}/`,
       };
+      renderedObject.Parameters.ResultConfiguration = Object.assign(renderedObject.Parameters.ResultConfiguration, customResultConfiguration);
     }
+
+    if (!renderedObject.Parameters.QueryExecutionContext.Catalog && !renderedObject.Parameters.QueryExecutionContext.Database) {
+      delete renderedObject.Parameters.QueryExecutionContext;
+    }
+
+    return { Resource: renderedObject.Resource, Parameters: sfn.FieldUtils.renderObject(renderedObject.Parameters) };
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:214]:
```suggestion
    return renderedObject;
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts [hunk 5]
```diff
@@ -153,4 +153,48 @@ describe('Start Query Execution', () => {
       },
     });
   });
+
+  // No preset execution context
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts:157]:
```suggestion
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts [hunk 6]
```diff
@@ -153,4 +153,48 @@ describe('Start Query Execution', () => {
...
+  test('missing executioncontext', () => {
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts:158]:
```suggestion
  test('omit QueryExecutionContext if no catalog or database name provided', () => {
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts [hunk 7]
```diff
@@ -153,4 +153,48 @@ describe('Start Query Execution', () => {
...
+      },
+      End: true,
+      Parameters: {
+        QueryString: 'CREATE DATABASE database',
+        ClientRequestToken: 'unique-client-request-token',
+        ResultConfiguration: {
+          OutputLocation: 's3://query-results-bucket/folder/',
+        },
+      },
+    });
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts:198]:
```suggestion
    expect(stack.resolve(task.toStateJson())).toMatchObject({
      Parameters: {
        QueryExecutionContext: ABSENT,
      },
    });
```

**Review by BenChaimberg — DISMISSED** [review]
Thank you so much for cleaning up our logic! A couple of further modifications to make things sleek

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts (hunk 4 above)
Comment (LukvonStrom) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:214]: I'll move the `sfn.FieldUtils.renderObject()` call to the object above then

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts (hunk 7 above)
Comment (LukvonStrom) [reply packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts:198]: Wow, that makes the test much clearer!

Comment (LukvonStrom) [reply packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts]: ABSENT only works with haveResource/haveResourceLike, but that makes the testcase much more verbose :/

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 8]
```diff
@@ -183,40 +183,26 @@ export class AthenaStartQueryExecution extends sfn.TaskStateBase {
...
-          },
-          ResultConfiguration: {
-            EncryptionConfiguration: this.renderEncryption(),
-          },
-          WorkGroup: this.props.workGroup,
-        }),
-      };
-    }
+    // see https://docs.aws.amazon.com/athena/latest/APIReference/API_StartQueryExecution.html
+
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:187]:
```suggestion
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 9]
```diff
@@ -183,40 +183,26 @@ export class AthenaStartQueryExecution extends sfn.TaskStateBase {
...
+    const renderedObject = {
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:188]:
```suggestion
    return {
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 10]
```diff
@@ -183,40 +183,26 @@ export class AthenaStartQueryExecution extends sfn.TaskStateBase {
...
+        } : undefined,
+        ResultConfiguration: {
+          EncryptionConfiguration: this.renderEncryption(),
+          OutputLocation: this.props.resultConfiguration?.outputLocation ? `s3://${this.props.resultConfiguration?.outputLocation?.bucketName}/${this.props.resultConfiguration?.outputLocation?.objectKey}/` : undefined,
+        },
+        WorkGroup: this.props?.workGroup,
+      }),
+    };
+
+    return renderedObject;
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:205]:
```suggestion
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 11]
```diff
@@ -183,40 +183,26 @@ export class AthenaStartQueryExecution extends sfn.TaskStateBase {
...
+      Parameters: sfn.FieldUtils.renderObject({
+        QueryString: this.props.queryString,
+        ClientRequestToken: this.props.clientRequestToken,
+        QueryExecutionContext: (this.props.queryExecutionContext?.catalogName || this.props.queryExecutionContext?.databaseName) ? {
+          Catalog: this.props.queryExecutionContext?.catalogName,
+          Database: this.props.queryExecutionContext?.databaseName,
+        } : undefined,
+        ResultConfiguration: {
+          EncryptionConfiguration: this.renderEncryption(),
+          OutputLocation: this.props.resultConfiguration?.outputLocation ? `s3://${this.props.resultConfiguration?.outputLocation?.bucketName}/${this.props.resultConfiguration?.outputLocation?.objectKey}/` : undefined,
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:199]:
```suggestion
          OutputLocation: this.props.resultConfiguration?.outputLocation ? `s3://${this.props.resultConfiguration.outputLocation.bucketName}/${this.props.resultConfiguration.outputLocation.objectKey}/` : undefined,
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/integ.start-query-execution.ts [hunk 12]
```diff
@@ -2,6 +2,7 @@ import * as sfn from '@aws-cdk/aws-stepfunctions';
 import * as cdk from '@aws-cdk/core';
 import { AthenaStartQueryExecution, EncryptionOption } from '../../lib';
 
+
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/integ.start-query-execution.ts:5]:
```suggestion
```

**Review by BenChaimberg — DISMISSED** [review]
awesome, looks great! just some minor stuff

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts [hunk 13]
```diff
@@ -153,4 +154,29 @@ describe('Start Query Execution', () => {
...
+
+
+    // THEN
+    expect(stack.resolve(task.toStateJson())).toMatchObject({
+      Parameters: {
+        QueryExecutionContext: ABSENT,
+      },
+    });
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts:180]: hm sorry, this doesn't seem to be doing what I want. can you change to the following to see if that works better?
```suggestion
    expect(stack.resolve(task.toStateJson())).not.toMatchObject({
      Parameters: {
        QueryExecutionContext: expect.anything(),
      },
    });
```

**Review by BenChaimberg — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts (hunk 13 above)
Comment (LukvonStrom) [reply packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts:180]: I just tested that - does not work. New approach is ` expect(stack.resolve(task.toStateJson())).not.toHaveProperty("Parameters.QueryExecutionContext")`
- testing that right now

Comment (LukvonStrom) [reply packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts:180]: I see my error - I missed the call to the actual `anything` method and used it as property.
But I think the solution via .not.toHaveProperty looks better

Comment (LukvonStrom) [reply packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts]: it works :)

**Review by BenChaimberg — APPROVED** [review]

**Pull request merged**